src/groth16/verifier.rs

```diff
@@ -58,147 +47,221 @@ pub fn verify_proof<'a, E: Engine>(
     // A * B + inputs * (-gamma) + C * (-delta) = alpha * beta
     // which allows us to do a single final exponentiation.
 
-    Ok(E::final_exponentiation(&E::miller_loop(
-        [
-            (&proof.a.prepare(), &proof.b.prepare()),
-            (&acc.into_affine().prepare(), &pvk.neg_gamma_g2),
-            (&proof.c.prepare(), &pvk.neg_delta_g2),
-        ]
-        .iter(),
-    ))
-    .unwrap()
-        == pvk.alpha_g1_beta_g2)
+    // Miller Loop for alpha * beta
+    let mut ml_a_b = E::Fqk::zero();
+    // Miller Loop for C * (-delta)
+    let mut ml_all = E::Fqk::zero();
+    // Miller Loop for inputs * (-gamma)
+    let mut ml_acc = E::Fqk::zero();
+
+    POOL.install(|| {
+        // Start the two independent miller loops
+        rayon::scope(|s| {
+            // - Thread 1: Calculate ML alpha * beta
+            let ml_a_b = &mut ml_a_b;
+            s.spawn(move |_| {
+                *ml_a_b = E::miller_loop(&[(&proof.a.prepare(), &proof.b.prepare())]);
+            });
+
+            // - Thread 2: Calculate ML c * (-gamma)
+            let ml_all = &mut ml_all;
+            s.spawn(move |_| *ml_all = E::miller_loop(&[(&proof.c.prepare(), &pvk.neg_delta_g2)]));
```

Comment:
The comment says `-gamma`, but the code says `neg_delta_g2`.

Reply:
typo, fixed